Provide mermaid diagram of relation between BIDS schema, specification and validators

yarikoptic opened the pull request:
This is based of the mermaid diagrams in the blog post we have: https://bids-website.readthedocs.io/en/latest/blog/2024/11/13/bids-validator-2.html and the goal is to give a visual depiction of relations between the components mentioned on this page and elsewhere.

- I think with this we can Close #625 

Here is a screenshot of preview.  I do agree that text is a bit small'ish, so we might want to prune it but it is quite usable and IMHO useful... May be there is a way to make it "full screen"?

<details>
<summary>version1: horizontal, harder to read, no zooming etc</summary> 

![image](https://github.com/user-attachments/assets/b4185ba0-159f-43f7-a3b6-89bd3e33bdbe)

</details>

<details>
<summary>version2: vertical</summary> 
 
![image](https://github.com/user-attachments/assets/0661de47-cfa9-43f6-ba54-bd7641eadffe)

</details>

version3: removed bids-apps and stats-models: 

![image](https://github.com/user-attachments/assets/c750b01a-a72b-4c1a-8c97-fd67cd753cbb)

TODOs 
- [ ] add into figure that schema.json is published to https://jsr.io/@bids/schema and that is what used by deno validator

yarikoptic committed:
Provide mermaid diagram of relation between BIDS schema, specification and validators

This is based of the mermaid diagrams in the blog post we have:
https://bids-website.readthedocs.io/en/latest/blog/2024/11/13/bids-validator-2.html
and the goal is to give a visual depiction of relations between the components
mentioned on this page and elsewhere.

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

Remi-Gau commented:
I would remove the legacy validator from that diagram.
Will simplify the whole thing.

Remi-Gau commented:
The preview on read the docs shows a mermaid syntax error.

This does not mean there is a real syntax error but that's the kind of problem when I was trying to use some of the more advanced mermaid features with mkdocs.

effigies commented:
It's not rendering here: https://bids-website--626.org.readthedocs.build/en/626/standards/index.html

yarikoptic committed:
Boost required mkdocs-material to require at least 9.6.3

yarikoptic commented:
- it is likely the requirement to have newer mkdocs-material ... learning about us using `pip-compile`... let's see if helps

- re legacy one: I think it might still be worth being there for a year or so since many could still "relate" to it and here it explicitly depicts it... so "not sure" about removal just yet.

yarikoptic commented:
FWIW: does render now.

effigies commented:
Looks like too much stuff going on. I can't read it at 100% zoom.

![image](https://github.com/user-attachments/assets/e8275605-69a0-45ab-a60c-b19c2608619a#gh-light-mode-only)
![image](https://github.com/user-attachments/assets/7571fa34-7487-481a-91e6-87fbf3ee4a78#gh-dark-mode-only)

yarikoptic committed:
ENH: add panzoom plugin to zoom/drag mermaid figure(s)

https://github.com/PLAYG0N/mkdocs-panzoom

yarikoptic commented:
@candleindark -- would you be so kind to see if you could tame the drawing so it is of better fidelity without zooming?

Meanwhile, I have added here use of panzoom plugin (https://playg0n.github.io/mkdocs-panzoom/) which should enable zoom'ing "Mermaid and D2 are included by default". Seems to work nicely.

candleindark commented:
> @candleindark -- would you be so kind to see if you could tame the drawing so it is of better fidelity without zooming?
> 
I think the problem stems having too many items in the graph rather the arrangement of the items in the graph, but I will make an attempt. It would be nice if there is some mermaid support that allows zooming in of a particular region of the diagram, like the behavior of the [prezi](https://docs.dandiarchive.org/21_models_integration/) app.

I don't think this problem is particular to us though. Many linkml generated diagrams suffer the same problem in presentation (For example, https://concepts.datalad.org/s/temporal/unreleased/).

candleindark commented:
I think if we can get any large mermaid diagram to display in full window as https://anvaka.github.io/panzoom/demo/index.html would be great.

The example is a live demo of https://github.com/anvaka/panzoom which is the base for https://playg0n.github.io/mkdocs-panzoom/

yarikoptic commented:
@candleindark I already added zoom/dragging here -- check preview at https://bids-website--626.org.readthedocs.build/en/626/standards/index.html . Do you propose something instead/on top of it?

candleindark commented:
> @candleindark I already added zoom/dragging here -- check preview at https://bids-website--626.org.readthedocs.build/en/626/standards/index.html . Do you propose something instead/on top of it?

In a way, yes. I do think the one I mentioned is more better, but it is not a Mkdocs extension. I don't know if it's worth the time/effort to get it working.

candleindark commented:
> > @candleindark I already added zoom/dragging here -- check preview at https://bids-website--626.org.readthedocs.build/en/626/standards/index.html . Do you propose something instead/on top of it?
> 
> In a way, yes. I do think the one I mentioned is more better, but it is not a Mkdocs extension. I don't know if it's worth the time/effort to get it working.

Nevermind, I just found out some a feature of the mkdocs extension you are using. I think it is good enough.

candleindark committed:
style: use top-to-bottom orientation and hierarchical layout

This adjustment will allow items in the
mermaid diagram to appear bigger in the
webpage because the page is more limited
in the horizontal dimension. Additionally,
the syntax of the mermaid diagram has been
updated to the latest syntax,
e.g. `flowchart` instead of `graph`.
The intents of the lines in the diagram
are those recommended/defaulted by the
mermaid live editor at
https://www.mermaidchart.com

candleindark commented:
@yarikoptic I was able to make the items in the diagram appear a bit larger and sent a [PR](https://github.com/yarikoptic/BIDS/pull/1) to your branch.

yarikoptic committed:
Minor: fixup indentations

yarikoptic commented on mkdocs.yml:
```diff
@@ -20,6 +20,8 @@ plugins:
       # toggle to true if you are in CD/CI environment
       on_error_fail: true
+  - panzoom:
+      full_screen: True # default False
```
well, there are icons in the right top corner of the figure -- I felt it should be enough.

candleindark replied:
The buttons on the right corner only allow you to reset and zoom to full screen. Pressing `Alt/Option` will allow you to pan by dragging your mouse and zoom by scrolling up and down, full screen or not.

This hidden feature is very useful if one wants to adjust the image size more precisely or look at only a particular region of a very big diagram.

yarikoptic replied:
there is also the question mark icon which would show that "hint" you want to make here permanently visible, right?

yarikoptic commented on docs/standards/index.md:
```diff
@@ -12,3 +12,76 @@ You can find further information for:
+        specification["specification html"]
+        schema.json@{ label: "<a href='https://bids-specification.readthedocs.io/en/v1.10.0/schema.json'>schema.json</a>" }
+    end
+    subgraph s3["<a href=https://bids-standard.github.io/stats-models/>stats-models standard</a>"]
+    end
+    subgraph s4["<a href=https://bids-standard.github.io/execution-spec/>bids-apps standard</a>"]
+    end
```
I think we should kick those out since do not relate to BIDS schema anyhow ATM.

```suggestion
```

yarikoptic committed:
Remove bids-apps and stats-models since unrelated to schema

yarikoptic commented:
@effigies please check it out now : https://bids-website--626.org.readthedocs.build/en/626/standards/index.html . I also updated screenshot in original description

Remi-Gau commented:
> * re legacy one: I think it might still be worth being there for a year or so since many could still "relate" to it and here it explicitly depicts it... so "not sure" about removal just yet.

we will forget to remove it in a year

it useless complexifes the figure that is already WAY too complex in my opinion

we have no intention of keeping maintaining the old validator, so let's not emphasize it

the original blog post gradually explained things and complexified the figure incrementally: I think that the landing page for 'standards' should have a simpler version and we can put the more exhaustive version somewhere else

effigies commented:
Just a thought sparked by @Remi-Gau's comment: what if this also was a blog post that came with explanations and is tagged with an author and date?

I haven't reviewed, but I feel like the editorial bar is lower for a blog post than a landing page.

Remi-Gau commented:
> Just a thought sparked by @Remi-Gau's comment: what if this also was a blog post that came with explanations and is tagged with an author and date?
> 
> I haven't reviewed, but I feel like the editorial bar is lower for a blog post than a landing page.

I like the idea and we can have the simplified version in the landing page and pointing to the blog post for anyone who wants the gory details

yarikoptic commented:
I think it is a good idea -- indeed could be a more descriptive blog post describing the situation, pointing to the blog post with more details on validators of the past.  Coolio, we will move into a blog post. Thanks @effigies and @Remi-Gau .  I will just move it to draft for now then

yarikoptic commented on docs/standards/index.md:
```diff
@@ -12,3 +12,70 @@ You can find further information for:
+    end
+    subgraph subGraph6["bids-validator ≥ v2.0"]
+        web@{ label: "<a href=\"https://bids-standard.github.io/bids-validator/\">web</a>" }
+        deno["deno"]
+        cli["cli"]
+    end
+    subgraph subGraph7["python-validator ≥ v1.14.7"]
+        python["python"]
+        library["library"]
+        cli3["cli"]
+    end
+
+    schema -. is interpreted by .-> bidsschematools
+    markdown -- is interpreted by --> mkdocs@{ label: "<a href=''>mkdocs</a>" }
+    bidsschematools -- provides MACROS for --> mkdocs
+    mkdocs -- renders HTML --> specification
+    schema -- is instance of --> bids-schema["<a href=https://bids-website.readthedocs.io/en/latest/standards/schema/index.html>BIDS Schema</a>"]
+    bids-schema -. is implemented by .-> deno & bidsschematools & python
+    bids-schema -. is reflected in .-> metaschema.json@{ label: "<a href='https://github.com/bids-standard/bids-specification/blob/master/src/metaschema.json'>metaschema.json</a><br/>JSON Schema" }
+    schema.json -- is instance of --> bids-schema & metaschema.json
+    bidsschematools -- compiles YAMLs as --> schema.json
+    regex --> Node.js
+    Node.js --> web1 & cli1
+    deno --> web & cli
+    python --> library & cli3
+    specification -. is interpreted by .-> regex
+    schema.json -. is interpreted by .-> deno
+    bidsschematools -- is used by --> python
+
+    markdown@{ shape: docs}
+    schema@{ shape: docs}
+    bidsschematools@{ shape: proc}
+    mkdocs@{ shape: proc}
+    deno@{ shape: subproc}
+    metaschema.json@{ shape: doc}
+    schema.json@{ shape: doc}
+    Node.js@{ shape: subproc}
+    web1@{ shape: rect}
+    web@{ shape: rect}
+    python@{ shape: subproc}
```
adding 2nd copy of schema.json as we have on npx:

```suggestion
---
config:
  layout: dagre
---
flowchart TB
 subgraph s6["<a href=https://github.com/bids-standard/bids-specification>bids-standard/bids-specification</a>"]
        markdown@{ label: "<a href='https://github.com/bids-standard/bids-specification/tree/v1.10.0/src'>src/</a><br>markdown" }
        bidsschematools@{ label: "<a href='https://github.com/bids-standard/bids-specification/tree/v1.10.0/tools/schemacode'>tools/schemacode/</a><br/>bidsschematools" }
        schema@{ label: "<a href='https://github.com/bids-standard/bids-specification/tree/v1.10.0/src/schema'>src/schema/</a><br>YAMLs" }
  end
 subgraph s2["<a href=https://bids-specification.readthedocs.io>bids-specification.readthedocs.io</a>"]
        specification["specification html"]
        schema.json@{ label: "<a href='https://bids-specification.readthedocs.io/en/v1.10.0/schema.json'>schema.json</a>" }
  end
 subgraph s_jsr["<a href=https://jsr.io/@bids/schema>https://jsr.io/@bids/schema</a>"]
        schema.json-jsr["schema.json"]
  end
 subgraph s5["<a href=https://bids.neuroimaging.io>bids.neuroimaging.io</a>"]
        s2
  end
 subgraph subGraph5["legacy-validator @ v1.15.1"]
        Node.js["Node.js"]
        regex["filename patterns"]
        web1@{ label: "<a href=\"https://bids-standard.github.io/legacy-validator/\">web</a>" }
        cli1["cli"]
  end
 subgraph subGraph6["bids-validator ≥ v2.0"]
        web@{ label: "<a href=\"https://bids-standard.github.io/bids-validator/\">web</a>" }
        deno["deno"]
        cli["cli"]
  end
 subgraph subGraph7["python-validator ≥ v1.14.7"]
        python["python"]
        library["library"]
        cli3["cli"]
  end
    schema -. is interpreted by .-> bidsschematools
    markdown -- is interpreted by --> mkdocs@{ label: "<a href=''>mkdocs</a>" }
    bidsschematools -- provides MACROS for --> mkdocs
    mkdocs -- renders HTML --> specification
    schema -- is instance of --> bids-schema@{ label: "<a href=\"https://bids-website.readthedocs.io/en/latest/standards/schema/index.html\">BIDS Schema</a>" }
    bids-schema -. is implemented by .-> deno & bidsschematools & python
    bids-schema -. is reflected in .-> metaschema.json@{ label: "<a href='https://github.com/bids-standard/bids-specification/blob/master/src/metaschema.json'>metaschema.json</a><br/>JSON Schema" }
    schema.json -- is instance of --> bids-schema & metaschema.json
    schema.json-jsr -- is instance of --> bids-schema & metaschema.json
    bidsschematools -- compiles YAMLs as --> schema.json & schema.json-jsr
    regex --> Node.js
    Node.js --> web1 & cli1
    deno --> web & cli
    python --> library & cli3
    specification -. is interpreted by .-> regex
    schema.json-jsr -. is interpreted by .-> deno
    bidsschematools -- is used by --> python
    markdown@{ shape: docs}
    bidsschematools@{ shape: proc}
    schema@{ shape: docs}
    schema.json@{ shape: doc}
    schema.json-jsr@{ shape: doc}
    Node.js@{ shape: subproc}
    web1@{ shape: rect}
    web@{ shape: rect}
    deno@{ shape: subproc}
    python@{ shape: subproc}
    mkdocs@{ shape: proc}
    bids-schema@{ shape: rect}
    metaschema.json@{ shape: doc}
```

yarikoptic committed:
Added 2nd copy of schema.json as shipped to npx and used by validator

yarikoptic commented:
Idea for a blog is neat but I do not think it is to materialize any time soon.  So I am making this PR  a candidate for merge - "perfect is enemy of the good" and I had a second look at the diagram and still find it useful to orient visitor in all the intricacies of dependencies, with URL hyperlinks coming quite handy.

candleindark commented:
> @candleindark could you provide a quick update to the diagram by stripping off historical aspects?

By historical aspect, do you mean taking away the component of `legacy-validator @ v1.15.1`? Is there anything else?

yarikoptic commented:
yeah -- remove prior versions, just depict how it all works now
